internal/storage/storage_handle.go

```diff
@@ -42,57 +38,30 @@ type storageClient struct {
 	client *storage.Client
 }
 
-type StorageClientConfig struct {
-	ClientProtocol      mountpkg.ClientProtocol
-	MaxConnsPerHost     int
-	MaxIdleConnsPerHost int
-	TokenSrc            oauth2.TokenSource
-	HttpClientTimeout   time.Duration
-	MaxRetryDuration    time.Duration
-	RetryMultiplier     float64
-	UserAgent           string
-}
-
 // NewStorageHandle returns the handle of Go storage client containing
 // customized http client. We can configure the http client using the
 // storageClientConfig parameter.
-func NewStorageHandle(ctx context.Context, clientConfig StorageClientConfig) (sh StorageHandle, err error) {
-	var transport *http.Transport
-	// Using http1 makes the client more performant.
-	if clientConfig.ClientProtocol == mountpkg.HTTP1 {
-		transport = &http.Transport{
-			MaxConnsPerHost:     clientConfig.MaxConnsPerHost,
-			MaxIdleConnsPerHost: clientConfig.MaxIdleConnsPerHost,
-			// This disables HTTP/2 in transport.
-			TLSNextProto: make(
-				map[string]func(string, *tls.Conn) http.RoundTripper,
-			),
-		}
-	} else {
-		// For http2, change in MaxConnsPerHost doesn't affect the performance.
-		transport = &http.Transport{
-			DisableKeepAlives: true,
-			MaxConnsPerHost:   clientConfig.MaxConnsPerHost,
-			ForceAttemptHTTP2: true,
+func NewStorageHandle(ctx context.Context, clientConfig storageutil.StorageClientConfig) (sh StorageHandle, err error) {
+
+	var clientOpts []option.ClientOption
+
+	// Add WithHttpClient option.
+	if clientConfig.ClientProtocol == mountpkg.HTTP1 || clientConfig.ClientProtocol == mountpkg.HTTP2 {
+		httpClient, clientErr := storageutil.CreateHttpClientObj(&clientConfig)
+		if clientErr != nil {
+			err = fmt.Errorf("while creating http endpoint: %w", clientErr)
+			return
 		}
-	}
 
-	// Custom http client for Go Client.
-	httpClient := &http.Client{
-		Transport: &oauth2.Transport{
-			Base:   transport,
-			Source: clientConfig.TokenSrc,
-		},
-		Timeout: clientConfig.HttpClientTimeout,
+		clientOpts = append(clientOpts, option.WithHTTPClient(httpClient))
 	}
 
-	// Setting UserAgent through RoundTripper middleware
-	httpClient.Transport = &userAgentRoundTripper{
-		wrapped:   httpClient.Transport,
-		UserAgent: clientConfig.UserAgent,
+	// Add Custom endpoint option.
+	if !storageutil.IsProdEndpoint(clientConfig.Endpoint) {
```

Comment:
what will happen if we always pass the endpoint?

Reply:
Ref: https://screenshot.googleplex.com/89HEJFnx2HtvSuz

Reply:
I am sorry. didnt follow.

Reply:
As per guidance by tritone@. We also need to take care of mtls endpoint if we set prod url WithEndpoint option.

Somewhere we override the MTLS endpoint with the provided endpoint - https://github.com/googleapis/google-api-go-client/blob/f7f1c7b4c610f6a41c130c3bb85d6a609e99fb37/internal/cba.go#L98